### ISSUE TITLE
Quick question: RLhF, three model on GPU?

### ISSUE THREAD

**Issue by vegasvv** [huggingface/trl#1245]
Hi, 

I am running your stacklamma RLHF script. Just a quick question, if I do full model training, do you put the reference model (SFT model) on CPU or GPU?

I ask this question because I wanna know/calculate the rough GPU memory you need if doing RLHF. I understand that the learning model is on GPU, the reward model is usually on GPU. From the PPO trainer code, it looks like you just create a reference model by deep copying of the learning  model, does this mean that it is on CPU "by-default"?

**Comment by hijkzzz** [comment]
Try OpenRLHF, which solves this problem perfectly using Ray: https://github.com/OpenLLMAI/OpenRLHF

**Comment by vegasvv** [comment]
@hijkzzz thanks! Is it easy to migrate over? I am trying to train 7B model full scale (no Lora) on A100

**Comment by hijkzzz** [comment]
> @hijkzzz thanks! Is it easy to migrate over? I am trying to train 7B model full scale (no Lora) on A100

Yes, the OpenRLHF is fully compatible with HF transformers.

**Comment by lvwerra** [comment]
Yes, all models are on GPU, although you can put the Reward model wherever you want as you have control over it outside the RLHF loop. If you want to save a bit more memory for the reference model you can share the first few layers between the reference and active model and only treat the last few layers separately. See [here](https://huggingface.co/docs/trl/v0.7.10/en/customization#memory-efficient-fine-tuning-by-sharing-layers).
